doc/DESIGN.md

```diff
@@ -283,3 +283,105 @@ const Key  *keyGetMeta(const Key *key, const char* metaName);
+
+On potential changes of the API/ABI as detected by
+[`elektra-icheck`](https://build.libelektra.org/job/elektra-icheck),
+please make sure the API has been reviewed according to the
+following 2 checklists:
+
+
+## Checklist for overall API
+
+### Consistency
+
+- [ ] Consistent naming schemes for enums, macros, typedefs and functions
+- [ ] Similar things are named similarly
+- [ ] Different things are named differently
+- [ ] The order of arguments should be consistent across similar functions
+
+### Structural Clarity
+
+- [ ] Functions with similar functionality have the same prefix
+
+### Compatibility
+
+- [ ] All Bindings have been updated to reflect the new API and work properly
+
+### Extensibility
+
+- [ ] New API is easily extensible with additional functionality
+- [ ] Components do not depend too heavily on each other
+
+
+## Checklist for each function
+
+### Documentation
+
+- [ ] Change is mentioned in the Compatibility section of the release notes
+- [ ] No inconsistencies between documentation and code
+- [ ] No inconsistencies between documentation and tests
+- [ ] Proper Documentation of thread-safety of function
+- [ ] All possible error states are documented
+- [ ] Configuration flags are documented
+
+#### Doxygen
+
+- [ ] Return Value
+- [ ] Precondition / Postcondition / Invariant
+- [ ] `@see`
+- [ ] `@since`
+- [ ] `@ingroup`
+
+### Naming
+
+- [ ] Abbreviations should be avoided in function names
+- [ ] Abbreviations should be avoided in parameter names
+- [ ] Parameter names should neither be too long, nor too short
+- [ ] Function names should neither be too long, nor too short
+- [ ] Function name should be clear and unambiguous
+- [ ] Parameter names should be clear and unambiguous
+
+### Compatibility
+
+- [ ] [Symbol versioning](/doc/dev/symbol-versioning.md)
+      is correct for breaking changes
+- [ ] ABI/API changes are forward-compatible (breaking backwards-compatibility
+      to add additional symbols is fine)
+
+### Parameter & Return Types
+
+- [ ] Functions should return the most specific type possible
+- [ ] Functions should require the most liberal type possible
+- [ ] Functions should use constant types instead of boolean types
+      wherever sensible
+- [ ] Wherever possible, function parameters should be `const`
+- [ ] Functions should have the least amount of parameters feasible
+
+### Structural Clarity
+
+- [ ] Function should do exactly one thing
+- [ ] Function should have no side effects
+- [ ] Configuration options are named constants
+- [ ] Function belongs to appropriate group
+
+### Memory Management
+
+- [ ] Memory Management should be handled by the function wherever possible
+- [ ] Functions, who require a large amount of memory to be allocated,
+      state so in their documentation
+
+### Extensibility
+
+- [ ] Function is easily extensible with additional functionality via bitmap-flags
+- [ ] Documentation does not impose limits, that would hinder further extensions
+
+### Tests
```

Comment:
I am still very undecided on this block. From my understanding it should be possible to judge the bullet points by just looking at the signature and the documentation of a function - no reading of source code whatsoever. For this part one would have to read all the existing tests for a given function, which would require a lot of effort compared to the rest.

Reply:
I am afraid that this step is unavoidable so that we can know that our tests (the real behavior) is consistent to the docu (the expected behavior). Usually a review of the whole implementation is needed to ensure this, we already make a shortcut by only looking at the tests.